### ISSUE TITLE
Object locks usage with sync-to command

### ISSUE THREAD

**Issue by RT2Code** [kopia/kopia#3759]
Hello,

I have a local repository that I synchronize daily to B2 with the sync-to command. I'm trying to enable object locks, but I'm unsure how to do it, or if it is even possible with the sync-to command.

I first tried to set the --retention-period and --retention-mode parameters to my local repository, but I got the following error :
`ERROR error updating repository parameters: error setting parameters: unable to write blobcfg blob: PutBlob() failed for "kopia.blobcfg": blob-retention: unsupported put-blob option`

It seems it is unsupported on local repositories, which makes sense, so I set these parameters on my remote B2 repository instead, and it worked. I then enabled object locks extension with the following command : `kopia maintenance set --extend-object-locks true`.

The issue is that, from what I get from the documentation, it's the maintenance step that is in charge of renewing the object locks retention dates, and maintenances are only done on my local repository, on which object locks aren't enabled, and therefore not updated. When I synchronize it to the remote B2 repository, the files are basically copied as is, and ends up on the B2 bucket without any object locks retention dates.

Am I getting this right? If yes, should I run another maintenance step on the B2 repository right after each synchronization?

**Comment by bbccdd** [comment]
**Local kopia repository synced to backblaze using their S3-api ransomware protection** is my wish as well. 

Same error though. But conceptually it does make sense to me to apply retention mode and periods to the local repo. Especially since retention mode is explicitly listed as command parameters of the [repository create **filesystem**](https://kopia.io/docs/reference/command-line/common/repository-create-filesystem/) command:
- _--retention-mode_ "Set the blob retention-mode for supported storage backends."
- _--retention-period_ "Set the blob retention-period for supported storage backends."

**Comment by bbccdd** [comment]
Or, should we do it the other way around? 

1. So set up s3 repo with ransomware protection ([as documented](https://kopia.io/docs/advanced/ransomware-protection)). 
2. Connect to that repo & start the kopia repository server ([as mentioned in the forum](https://kopia.discourse.group/t/server-with-remote-storage/1705/4?u=bbccdd)).
3. And finally and sync back from remote to local repo?

Result would be similar albeit less performant. 

Just thinking out loud here.

**Comment by bbccdd** [comment]
> Or, should we do it the other way around?
> 
> 1. So set up s3 repo with ransomware protection ([as documented](https://kopia.io/docs/advanced/ransomware-protection)).
> 2. Connect to that repo & start the kopia repository server ([as mentioned in the forum](https://kopia.discourse.group/t/server-with-remote-storage/1705/4?u=bbccdd)).
> 3. And finally and sync back from remote to local repo?
> 
> Result would be similar albeit less performant.
> 
> Just thinking out loud here.

This works just fine. Tested it.

Funny: After having 'synce-to filesystem' from the remote S3 storage location, I can't delete snapshots (within the retention/lock time). This - in my simple mind - implies that there is some awareness of retention periods in local repo's.

`kopia@host:~$ kopia snapshot delete --all-snapshots-for-source --delete "root@pihole:/home"
Deleting snapshot ca43ec7d53a2a8e9c0ffbe2976be87fe of root@pihole:/home at 2024-04-02 13:46:27 CEST...ERROR error flushing writer: error flushing contents: error writing pending content: error writing pack content: error writing pack: can't save pack data blob qd05ee40e72faacc21e5f423288dcda76-s050b0a6b1caf5316127: error writing pack file: **blob-retention: unsupported put-blob option**`

Again, this setup is not ideal for performance reasons, but it works!

**Comment by RT2Code** [comment]
I thought about doing it this way too, but my internet connection is really bad. As you said, synchronizing from local to remote makes more sense because it only requires uploading the files, while the other way around requires downloading them too.

**Comment by bbccdd** [comment]
Tried the reverse of the reverse: 

1. created repo on s3, 
2. started repository server, 
3. made a few snapshots, 
4. synced that to filesystem, 
5. stopped repo server, 
6. connected to filesystem repo, 
7. started repo server on that and 
8. made new snapshots (with the idea that I could than again 'sync-to' s3 having created the repo there.). 
9. But it failed as well.
`"Error: error flushing writer: error flushing contents: error writing pending content: error writing pack content: error writing pack: can't save pack data blob p023269f7d9c3baaf74d82adb399a31c2-s1a38892142748989127: error writing pack file: blob-retention: unsupported put-blob option."`

**Comment by bbccdd** [comment]
TL;DR it is **not possible to use a local filesystem based repository** and sync that to an S3 storage location while using **object locking**, unfortunately, for now.

Error: "error writing pack file: blob-retention: unsupported put-blob option."

**Comment by fastlorenzo** [comment]
I made a first attempt at fixing it within Kopia directly; if somebody wants to try it out, you can check PR #5508 
It would indeed be great to be able to automatically create a 3rd immutable copy from a local or SFTP repository.
